hudi-common/src/main/java/org/apache/hudi/avro/AvroSchemaUtils.java

```diff
@@ -428,25 +408,95 @@ public static void checkSchemaCompatible(
       boolean allowProjection,
       Set<String> dropPartitionColNames) throws SchemaCompatibilityException {
 
-    String errorMessage = null;
-
-    if (!allowProjection && !canProject(tableSchema, writerSchema, dropPartitionColNames)) {
-      errorMessage = "Column dropping is not allowed";
+    if (!allowProjection) {
+      List<Schema.Field> missingFields = findMissingFields(tableSchema, writerSchema, dropPartitionColNames);
+      if (!missingFields.isEmpty()) {
+        throw new MissingSchemaFieldException(missingFields.stream().map(Schema.Field::name).collect(Collectors.toList()));
+      }
     }
 
     // TODO(HUDI-4772) re-enable validations in case partition columns
     //                 being dropped from the data-file after fixing the write schema
-    if (dropPartitionColNames.isEmpty() && shouldValidate && !isSchemaCompatible(tableSchema, writerSchema)) {
-      errorMessage = "Failed schema compatibility check";
+    if (dropPartitionColNames.isEmpty() && shouldValidate) {
+      AvroSchemaCompatibility.SchemaPairCompatibility result =
+          AvroSchemaCompatibility.checkReaderWriterCompatibility(writerSchema, tableSchema, true);
+      if (result.getType() != AvroSchemaCompatibility.SchemaCompatibilityType.COMPATIBLE) {
+        throw new SchemaBackwardsCompatibilityException(result);
+      }
     }
+  }
 
-    if (errorMessage != null) {
-      String errorDetails = String.format(
-          "%s\nwriterSchema: %s\ntableSchema: %s",
-          errorMessage,
-          writerSchema,
-          tableSchema);
-      throw new SchemaCompatibilityException(errorDetails);
+  /**
+   * Validate whether the {@code incomingSchema} is a valid evolution of {@code tableSchema}.
+   *
+   * @param incomingSchema schema of the incoming dataset
+   * @param tableSchema latest table schema
+   */
+  public static void checkValidEvolution(Schema incomingSchema, Schema tableSchema) {
```

Comment:
-1 Schemas can be quite large. We should just log out the relevant context in my opinion so the user can quickly see what is wrong instead of requiring them to manually diff the two.

Reply:
sounds good. I am not too strong on that. I am good w/ cur state.

Reply:
pinpointing mismatching field(s) would be more helpful. Today we copy both the schemas from the log and take diff anyway, so why not just log the diff itself.

Reply:
We print the schemas if incompatible before this PR.  I still suggest printing out the schema.  I assume this only happens once per batch, not per record, so it's ok.

Reply:
Adding to that, my point is we should print both the specific context of what's causing the incompatibility, and both schemas (writer schema, and table schema).  Former helps user to check the specific issue, and later helps users see the whole schema if the user could not understand the specific issue, and devs to check the bug in schema incompatibility check if that happens.